### PR TITLE
Feature costs to consumptions

### PR DESCRIPTION
**Pull request by Winston-503**
- Add dollar costs into `LLMResult.consumptions` based on cost cards of popular models
- Implement cost calculations for Anthropic and OpenAI prompt caching
- Add Gemini token counts
- Add consumptions logging for `LLMLoggingMiddleware` and `LLMFileLoggingMiddleware`
- Fix typo in `LLMMessageTokenCounterBase`

### CODE REVIEW

File: council/llm/llm_base.py [hunk 1]
```diff
@@ -113,3 +113,100 @@ def post_chat_request(self, context: LLMContext, messages: Sequence[LLMMessage],
...
+            "prompt",
+            "completion",
+            "total",
+            "reasoning",  # OpenAI o1
+            "cache_creation_prompt",  # Anthropic prompt caching
+            "cache_read_prompt",  # Anthropic & OpenAI prompt caching
+        ]
+        result = f"{self.model}:_"
+        if token_kind not in options:
+            raise ValueError(f"Unknown kind for LLMConsumptionCalculator; expected one of `{','.join(options)}`")
```
Comment (aflament) [inline council/llm/llm_base.py:165]:
```suggestion
            raise ValueError(f"`{token_kind}` Unknown kind for LLMConsumptionCalculator; expected one of `{','.join(options)}`")
```

File: council/llm/llm_base.py [hunk 2]
```diff
@@ -113,3 +113,100 @@ def post_chat_request(self, context: LLMContext, messages: Sequence[LLMMessage],
     @abc.abstractmethod
     def _post_chat_request(self, context: LLMContext, messages: Sequence[LLMMessage], **kwargs: Any) -> LLMResult:
         pass
+
+
+class LLMCostCard:
```
Comment (aflament) [inline council/llm/llm_base.py:118]: move `LLMCostCard` and `LLMConsumptionCalculator` to a `llm_cost.py` file

File: council/llm/llm_base.py [hunk 3]
```diff
@@ -113,3 +113,100 @@ def post_chat_request(self, context: LLMContext, messages: Sequence[LLMMessage],
...
+
+    def output_cost(self, tokens: int) -> float:
+        return tokens * self.output / 1e6
+
+    def get_costs(self, prompt_tokens: int, completion_tokens: int) -> Tuple[float, float]:
+        """Return tuple of (prompt_tokens_cost, completion_token_cost)"""
+        return self.input_cost(prompt_tokens), self.output_cost(completion_tokens)
+
+
+class LLMConsumptionCalculator(abc.ABC):
```
Comment (aflament) [inline council/llm/llm_base.py:147]:
```suggestion
class LLMConsumptionCalculatorBase(abc.ABC):
```

File: council/llm/llm_message.py [hunk 4]
```diff
@@ -249,7 +249,7 @@ def normalize(self) -> str:
         return normalized_content
 
 
-class LLMessageTokenCounterBase(abc.ABC):
+class LLMMessageTokenCounterBase(abc.ABC):
```
Comment (aflament) [inline council/llm/llm_message.py:252]: Good catch 👍

File: council/llm/anthropic_llm.py [hunk 5]
```diff
@@ -32,6 +34,88 @@ def count_messages_token(self, messages: Sequence[LLMMessage]) -> int:
         return tokens
 
 
+class AnthropicConsumptionCalculator(LLMConsumptionCalculator):
+    # https://www.anthropic.com/pricing#anthropic-api
+    COSTS: Mapping[str, LLMCostCard] = {
```
[the remark this answers is not in the document]
Comment (Winston-503) [reply council/llm/anthropic_llm.py:39]: Was planning to do in a follow-up

File: council/llm/anthropic_llm.py [hunk 6]
```diff
@@ -32,6 +34,88 @@ def count_messages_token(self, messages: Sequence[LLMMessage]) -> int:
...
+            - cache_creation_prompt, cache_read_prompt, prompt, completion and total tokens
+            - costs if both regular and caching LLMCostCards can be found
+        """
+        consumptions = self.get_cache_token_consumptions(usage) + self.get_cache_cost_consumptions(usage)
+        return self.filter_zeros(consumptions)  # could occur for cache tokens
+
+    def get_cache_token_consumptions(self, usage: Dict[str, int]) -> List[Consumption]:
+        total = sum(
+            [
+                usage["cache_creation_prompt_tokens"],
```
[the remark this answers is not in the document]
Comment (Winston-503) [reply council/llm/anthropic_llm.py:74]: Was planning to do in a follow-up

**Review by aflament — APPROVED** [review]

**Pull request merged**